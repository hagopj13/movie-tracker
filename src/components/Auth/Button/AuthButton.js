// @flow
import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { makeStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';

import DialogTypes from 'components/Dialog/types';
import authSelectors from 'store/auth/auth.selectors';
import authActions from 'store/auth/auth.actions';
import AuthActionTypes from 'store/auth/auth.types';
import loadingSelectors from 'store/api/loading/loading.selectors';
import dialogActions from 'store/ui/dialog/dialog.actions';

type Props = {
  isAuth: boolean,
  isLogoutLoading: boolean,
  onLogout: () => void,
  onShowLoginDialog: () => void,
};

const useStyles = makeStyles((theme) => ({
  root: {
    minWidth: 72,
    textTransform: 'none',
    fontSize: theme.typography.h6.fontSize,
    fontWeight: theme.typography.fontWeightRegular,
    '&$disabled': {
      color: theme.palette.grey[300],
    },
  },
  disabled: {},
  logout: {
    color: theme.palette.common.white,
  },
}));

const AuthButton = (props: Props) => {
  const { isAuth, isLogoutLoading, onLogout, onShowLoginDialog } = props;
  const classes = useStyles();

  return (
    <>
      {isAuth ? (
        <Button
          classes={{ root: classes.root, disabled: classes.disabled, text: classes.logout }}
          disabled={isLogoutLoading}
          onClick={onLogout}
        >
          Logout
        </Button>
      ) : (
        <Button className={classes.root} color="primary" onClick={onShowLoginDialog}>
          Login
        </Button>
      )}
    </>
  );
};

const mapStateToProps = createStructuredSelector({
  isAuth: authSelectors.selectIsAuth,
  isLogoutLoading: loadingSelectors.createIsLoadingSelector([AuthActionTypes.LOGOUT]),
});

const mapDispatchToProps = {
  onLogout: authActions.logout,
  onShowLoginDialog: () => dialogActions.showDialog(DialogTypes.LOGIN),
};

export default connect(mapStateToProps, mapDispatchToProps)(AuthButton);
