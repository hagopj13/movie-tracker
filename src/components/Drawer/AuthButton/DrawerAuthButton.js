// @flow
import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import LogoutIcon from '@material-ui/icons/Lock';
import LoginIcon from '@material-ui/icons/LockOpen';

import dialogTypes from 'config/dialogTypes';
import authSelectors from 'store/auth/auth.selectors';
import authActions from 'store/auth/auth.actions';
import dialogActions from 'store/ui/dialog/dialog.actions';

type Props = {
  isAuth: boolean,
  onLogout: () => void,
  onShowLoginDialog: () => void,
};

export const DrawerAuthButton = (props: Props) => {
  const { isAuth, onLogout, onShowLoginDialog } = props;

  return (
    <List>
      {isAuth ? (
        <ListItem button onClick={onLogout}>
          <ListItemIcon>
            <LogoutIcon />
          </ListItemIcon>
          <ListItemText primary="Logout" />
        </ListItem>
      ) : (
        <ListItem button onClick={onShowLoginDialog}>
          <ListItemIcon>
            <LoginIcon />
          </ListItemIcon>
          <ListItemText primary="Login" />
        </ListItem>
      )}
    </List>
  );
};

const mapStateToProps = createStructuredSelector({
  isAuth: authSelectors.selectIsAuth,
});

const mapDispatchToProps = {
  onLogout: authActions.logout,
  onShowLoginDialog: () => dialogActions.showDialog(dialogTypes.LOGIN),
};

export default connect(mapStateToProps, mapDispatchToProps)(DrawerAuthButton);
