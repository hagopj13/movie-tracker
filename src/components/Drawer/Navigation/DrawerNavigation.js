// @flow
import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { useLocation, NavLink } from 'react-router-dom';
import MenuList from '@material-ui/core/MenuList';
import MenuItem from '@material-ui/core/MenuItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';

import authSelectors from 'store/auth/auth.selectors';
import navigationItems from 'config/navigationItems';

type Props = {
  isAuth: boolean,
};

export const DrawerNavigation = (props: Props) => {
  const { isAuth } = props;

  const location = useLocation();

  const isActiveRoute = (path: string) => location.pathname === path;

  return (
    <MenuList>
      {navigationItems.map(
        (navItem) =>
          (isAuth || !navItem.requiresAuth) && (
            <NavLink
              key={navItem.id}
              to={navItem.to}
              style={{ textDecoration: 'none', color: 'inherit' }}
            >
              <MenuItem selected={isActiveRoute(navItem.to)}>
                <ListItemIcon>{navItem.icon}</ListItemIcon>
                <ListItemText primary={navItem.label} />
              </MenuItem>
            </NavLink>
          ),
      )}
    </MenuList>
  );
};

const mapStateToProps = createStructuredSelector({
  isAuth: authSelectors.selectIsAuth,
});

export default connect(mapStateToProps)(DrawerNavigation);
