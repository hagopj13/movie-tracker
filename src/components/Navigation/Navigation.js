// @flow
import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { NavLink } from 'react-router-dom';
import { makeStyles } from '@material-ui/core/styles';
import Link from '@material-ui/core/Link';

import authSelectors from 'store/auth/auth.selectors';

type Props = {
  isAuth: boolean,
  className: string,
};

const useStyles = makeStyles((theme) => ({
  navLink: {
    color: theme.palette.common.white,
    marginRight: theme.spacing(2),
    '&:last-child': {
      marginRight: 0,
    },
  },
  selectedNavLink: {
    color: theme.palette.primary.main,
  },
}));

const Navigation = (props: Props) => {
  const { isAuth, className } = props;

  const classes = useStyles();

  return (
    <div className={className}>
      <Link
        component={NavLink}
        exact
        to="/"
        className={classes.navLink}
        activeClassName={classes.selectedNavLink}
      >
        Discover
      </Link>
      <Link
        component={NavLink}
        exact
        to="/upcoming"
        className={classes.navLink}
        activeClassName={classes.selectedNavLink}
      >
        Upcoming
      </Link>
      {isAuth && (
        <Link
          component={NavLink}
          to="/profile"
          className={classes.navLink}
          activeClassName={classes.selectedNavLink}
        >
          Profile
        </Link>
      )}
    </div>
  );
};

const mapStateToProps = createStructuredSelector({
  isAuth: authSelectors.selectIsAuth,
});

export default connect(mapStateToProps)(Navigation);
