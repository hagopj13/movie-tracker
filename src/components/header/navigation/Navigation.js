import React from 'react';
import { NavLink } from 'react-router-dom';

import './navigation.scss';

const Navigation = () => (
  <div className="navigation">
    <NavLink exact to="/" className="nav-link" activeClassName="nav-link-selected">
      Discover
    </NavLink>
    <NavLink to="/profile" className="nav-link" activeClassName="nav-link-selected">
      Profile
    </NavLink>
  </div>
);

export default Navigation;
