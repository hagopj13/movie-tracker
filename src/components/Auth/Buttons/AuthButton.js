// @flow
import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { makeStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';

import DialogTypes from 'components/UI/Dialog/types';
import { selectIsAuth } from 'store/auth/auth.selectors';
import { logout } from 'store/auth/auth.actions';
import AuthActionTypes from 'store/auth/auth.types';
import { createIsLoadingSelector } from 'store/api/loading/loading.selectors';
import { showDialog } from 'store/ui/dialog/dialog.actions';

type Props = {
  isAuth: boolean,
  isLogoutLoading: boolean,
  onLogout: () => void,
  onShowLoginDialog: () => void,
};

const useStyles = makeStyles((theme) => ({
  root: {
    minWidth: 72,
    textTransform: 'none',
    fontSize: 14,
    '&$disabled': {
      color: theme.palette.grey[300],
    },
  },
  disabled: {},
  logout: {
    color: theme.palette.common.white,
  },
}));

const AuthButtons = (props: Props) => {
  const { isAuth, isLogoutLoading, onLogout, onShowLoginDialog } = props;
  const classes = useStyles();

  return (
    <>
      {isAuth ? (
        <Button
          classes={{ root: classes.root, disabled: classes.disabled, text: classes.logout }}
          disabled={isLogoutLoading}
          onClick={onLogout}
        >
          Logout
        </Button>
      ) : (
        <Button className={classes.root} color="primary" onClick={onShowLoginDialog}>
          Login
        </Button>
      )}
    </>
  );
};

const mapStateToProps = createStructuredSelector({
  isAuth: selectIsAuth,
  isLogoutLoading: createIsLoadingSelector([AuthActionTypes.LOGOUT]),
});

const mapDispatchToProps = {
  onLogout: logout,
  onShowLoginDialog: () => showDialog(DialogTypes.LOGIN),
};

export default connect(mapStateToProps, mapDispatchToProps)(AuthButtons);
