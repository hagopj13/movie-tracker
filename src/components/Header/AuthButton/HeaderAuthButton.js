// @flow
import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { makeStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';

import dialogTypes from 'config/dialogTypes';
import authSelectors from 'store/auth/auth.selectors';
import authActions from 'store/auth/auth.actions';
import AuthActionTypes from 'store/auth/auth.types';
import loadingSelectors from 'store/api/loading/loading.selectors';
import dialogActions from 'store/ui/dialog/dialog.actions';

type Props = {
  isAuth: boolean,
  isLogoutLoading: boolean,
  onLogout: () => void,
  onShowLoginDialog: () => void,
};

const useStyles = makeStyles((theme) => ({
  root: {
    minWidth: 72,
    textTransform: 'none',
    fontSize: theme.typography.h6.fontSize,
    '&$disabled': {
      color: theme.palette.grey[300],
    },
  },
  disabled: {},
  logoutText: {
    color: theme.palette.common.white,
  },
}));

export const HeaderAuthButton = (props: Props) => {
  const { isAuth, isLogoutLoading, onLogout, onShowLoginDialog } = props;
  const classes = useStyles();

  return isAuth ? (
    <Button
      classes={{ root: classes.root, disabled: classes.disabled, text: classes.logoutText }}
      disabled={isLogoutLoading}
      onClick={onLogout}
    >
      Logout
    </Button>
  ) : (
    <Button className={classes.root} color="primary" onClick={onShowLoginDialog}>
      Login
    </Button>
  );
};

const mapStateToProps = createStructuredSelector({
  isAuth: authSelectors.selectIsAuth,
  isLogoutLoading: loadingSelectors.createIsLoadingSelector([AuthActionTypes.LOGOUT]),
});

const mapDispatchToProps = {
  onLogout: authActions.logout,
  onShowLoginDialog: () => dialogActions.showDialog(dialogTypes.LOGIN),
};

export default connect(mapStateToProps, mapDispatchToProps)(HeaderAuthButton);
