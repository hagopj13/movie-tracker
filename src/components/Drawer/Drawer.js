// @flow
import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import MuiDrawer from '@material-ui/core/Drawer';
import IconButton from '@material-ui/core/IconButton';
import Divider from '@material-ui/core/Divider';
import ChevronLeftIcon from '@material-ui/icons/ChevronLeft';

import Navigation from './Navigation/DrawerNavigation';
import AuthButton from './AuthButton/DrawerAuthButton';

type Props = {
  isOpen: boolean,
  onClose: () => void,
};

const DRAWER_WIDTH = 240;

const useStyles = makeStyles((theme) => ({
  drawerPaper: {
    width: DRAWER_WIDTH,
  },
  drawerHeader: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'flex-end',
    padding: theme.spacing(0, 1),
    minHeight: 64,
  },
}));

const Drawer = (props: Props) => {
  const { isOpen, onClose } = props;

  const classes = useStyles();

  return (
    <MuiDrawer
      anchor="left"
      open={isOpen}
      onClose={onClose}
      classes={{
        paper: classes.drawerPaper,
      }}
    >
      <div className={classes.drawerHeader}>
        <IconButton onClick={onClose}>
          <ChevronLeftIcon />
        </IconButton>
      </div>
      <Divider />
      <div role="presentation" onClick={onClose}>
        <Navigation />
      </div>
      <Divider />
      <div role="presentation" onClick={onClose}>
        <AuthButton />
      </div>
    </MuiDrawer>
  );
};

export default Drawer;
