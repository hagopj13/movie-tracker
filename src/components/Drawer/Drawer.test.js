import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';
import MuiDrawer from '@material-ui/core/Drawer';
import IconButton from '@material-ui/core/IconButton';

import Drawer from './Drawer';
import Navigation from './Navigation/DrawerNavigation';
import AuthButton from './AuthButton/DrawerAuthButton';

describe('Drawer component', () => {
  let shallow;
  let isOpen;
  let wrapper;

  const mockClose = jest.fn();

  beforeAll(() => {
    shallow = createShallow();
  });

  beforeEach(() => {
    jest.clearAllMocks();
    isOpen = true;
    const mockProps = {
      isOpen,
      onClose: mockClose,
    };
    wrapper = shallow(<Drawer {...mockProps} />);
  });

  it('should render the Drawer component correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should pass isOpen as a prop to MuiDrawer', () => {
    expect(wrapper.find(MuiDrawer).prop('open')).toBe(isOpen);
  });

  it('should call the onClose prop when close icon is clicked', () => {
    wrapper.find(IconButton).simulate('click');
    expect(mockClose).toHaveBeenCalledTimes(1);
  });

  it('should call the onClose prop when a click occurs in the navigation div', () => {
    wrapper.find(Navigation).parent().simulate('click');
    expect(mockClose).toHaveBeenCalledTimes(1);
  });

  it('should call the onClose prop when a click occurs in the auth button div', () => {
    wrapper.find(AuthButton).parent().simulate('click');
    expect(mockClose).toHaveBeenCalledTimes(1);
  });
});
