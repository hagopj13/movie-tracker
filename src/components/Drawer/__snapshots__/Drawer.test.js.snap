// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer component should render the Drawer component correctly 1`] = `
<WithStyles(ForwardRef(Drawer))
  anchor="left"
  classes={
    Object {
      "paper": "makeStyles-drawerPaper-1",
    }
  }
  onClose={[MockFunction]}
  open={true}
>
  <div
    className="makeStyles-drawerHeader-2"
  >
    <WithStyles(ForwardRef(IconButton))
      onClick={[MockFunction]}
    >
      <ChevronLeftIcon />
    </WithStyles(ForwardRef(IconButton))>
  </div>
  <WithStyles(ForwardRef(Divider)) />
  <div
    onClick={[MockFunction]}
    role="presentation"
  >
    <Connect(DrawerNavigation) />
  </div>
  <WithStyles(ForwardRef(Divider)) />
  <div
    onClick={[MockFunction]}
    role="presentation"
  >
    <Connect(DrawerAuthButton) />
  </div>
</WithStyles(ForwardRef(Drawer))>
`;
